File: isabella.ml/src/canon/confidential_balance.ml
(** Confidential balance proof helpers
    Derived from the verified Canon/ZK/Confidential_Balance theory.

    This module keeps the SDK surface native and deterministic while remaining
    a thin composition layer over Isabelle-exported Canon modules. *)

type scalar_commit_params = Commit_sis.commit_params
type commit_opening = Commit_sis.commit_opening

type balance_proof = {
  balance_as : int list list;
  balance_zs : int list list;
}

let make_balance_proof balance_as balance_zs = { balance_as; balance_zs }

let balance_fs_rounds = Repeated_fs.fixed_fs_rounds

let make_scalar_commit_params m n2 q beta =
  Commit_sis.make_commit_params 1 n2 m q beta

let valid_scalar_commit_params p =
  Commit_sis.valid_commit_params p && p.Commit_sis.cp_n1 = 1

let valid_confidential_commit_key p ck =
  Commit_sis.separating_commit_key p ck

let rand_commit_key p ck =
  List.map (fun row -> snd (Listvec.split_vec p.Commit_sis.cp_n1 row)) ck

let rand_commit p ck r =
  Zq.vec_mod (Listvec.mat_vec_mult (rand_commit_key p ck) r) p.Commit_sis.cp_q

let amount_of_opening op =
  match op.Commit_sis.open_msg with
  | x :: _ -> x
  | [] -> 0

let opening_add op1 op2 =
  Commit_sis.make_opening
    (Listvec.vec_add op1.Commit_sis.open_msg op2.Commit_sis.open_msg)
    (Listvec.vec_add op1.Commit_sis.open_rand op2.Commit_sis.open_rand)

let opening_sub op1 op2 =
  Commit_sis.make_opening
    (Listvec.vec_sub op1.Commit_sis.open_msg op2.Commit_sis.open_msg)
    (Listvec.vec_sub op1.Commit_sis.open_rand op2.Commit_sis.open_rand)

let aggregate_opening op_in1 op_in2 op_out1 op_out2 =
  opening_sub (opening_add op_in1 op_in2) (opening_add op_out1 op_out2)

let aggregate_randomness op_in1 op_in2 op_out1 op_out2 =
  (aggregate_opening op_in1 op_in2 op_out1 op_out2).Commit_sis.open_rand

let balance_commitment c_in1 c_in2 c_out1 c_out2 q =
  Zq.vec_mod
    (Listvec.vec_sub (Listvec.vec_add c_in1 c_in2) (Listvec.vec_add c_out1 c_out2))
    q

let valid_balance_witness p r =
  Listvec.valid_vec p.Commit_sis.cp_n2 r &&
  Norms.all_bounded r (4 * p.Commit_sis.cp_beta)

let valid_balance_mask p gamma y =
  Listvec.valid_vec p.Commit_sis.cp_n2 y && Norms.all_bounded y gamma

let balance_response_bound p gamma challenge =
  gamma + (abs challenge * (4 * p.Commit_sis.cp_beta))

let valid_balance_challenge p challenge =
  valid_scalar_commit_params p &&
  (challenge = 0 || challenge = 1)

let valid_balance_response p gamma challenge z =
  Listvec.valid_vec p.Commit_sis.cp_n2 z &&
  Norms.all_bounded z (balance_response_bound p gamma challenge)

let balance_relation p ck c r =
  valid_scalar_commit_params p &&
  valid_confidential_commit_key p ck &&
  valid_balance_witness p r &&
  rand_commit p ck r = c

let balance_sigma_commit = rand_commit

let balance_sigma_respond r y challenge =
  Listvec.vec_add y (Listvec.scalar_mult challenge r)

let sum_list = List.fold_left ( + ) 0

let canonical_balance_challenge _p ck c a =
  (sum_list (List.concat ck) + sum_list c + sum_list a) mod 2

let balance_sigma_verify p gamma ck c a challenge z =
  valid_scalar_commit_params p &&
  valid_confidential_commit_key p ck &&
  Listvec.valid_vec p.Commit_sis.cp_m a &&
  valid_balance_challenge p challenge &&
  valid_balance_response p gamma challenge z &&
  rand_commit p ck z =
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) p.Commit_sis.cp_q

let balance_fs_challenges _p ck c as_ =
  Repeated_fs.bool_fs_challenges
    (sum_list (List.concat ck) + sum_list c + sum_list (List.concat as_))

let balance_sigma_responds r ys es =
  Repeated_fs.sigma_response_rounds balance_sigma_respond r ys es

let balance_fs_prove p gamma ck c r ys =
  let as_ = List.map (balance_sigma_commit p ck) ys in
  let es = balance_fs_challenges p ck c as_ in
  let zs = balance_sigma_responds r ys es in
  if balance_relation p ck c r &&
     List.length ys = balance_fs_rounds &&
     List.for_all (valid_balance_mask p gamma) ys &&
     List.for_all2 (fun e z -> valid_balance_response p gamma e z) es zs
  then Some { balance_as = as_; balance_zs = zs }
  else None

let balance_fs_verify p gamma ck c proof =
  let as_ = proof.balance_as in
  let es = balance_fs_challenges p ck c as_ in
  let zs = proof.balance_zs in
  List.length as_ = balance_fs_rounds &&
  List.length zs = balance_fs_rounds &&
  List.for_all2
    (fun a (e, z) -> balance_sigma_verify p gamma ck c a e z)
    as_
    (List.combine es zs)

File: isabella.hs/src/Canon/ZK/Confidential_Balance.hs
-- | Canon.ZK.Confidential_Balance: native Haskell facade for the
-- confidential-balance proof slice exported from Canon.
module Canon.ZK.Confidential_Balance
  ( BalanceProof(..)
  , makeBalanceProof
  , balanceFsRounds
  , makeScalarCommitParams
  , validScalarCommitParams
  , validConfidentialCommitKey
  , randCommitKey
  , randCommit
  , amountOfOpening
  , balanceCommitment
  , aggregateRandomness
  , validBalanceWitness
  , validBalanceMask
  , balanceResponseBound
  , validBalanceChallenge
  , validBalanceResponse
  , balanceRelation
  , balanceSigmaCommit
  , balanceSigmaRespond
  , canonicalBalanceChallenge
  , balanceSigmaVerify
  , balanceFsChallenges
  , balanceSigmaResponds
  , balanceFsProve
  , balanceFsVerify
  ) where

import qualified Canon.Commit_sis as Commit
import qualified Canon.Listvec as Listvec
import qualified Canon.Norms as Norms
import qualified Canon.ZK.Internal.RepeatedFS as RepeatedFS
import qualified Canon.Zq as Zq

data BalanceProof = BalanceProof
  { balance_as :: [[Int]]
  , balance_zs :: [[Int]]
  }
  deriving (Eq, Show)

makeBalanceProof :: [[Int]] -> [[Int]] -> BalanceProof
makeBalanceProof = BalanceProof

balanceFsRounds :: Int
balanceFsRounds = RepeatedFS.fixedFsRounds

makeScalarCommitParams :: Int -> Int -> Int -> Int -> Commit.CommitParams
makeScalarCommitParams m n2 q beta = Commit.makeCommitParams 1 n2 m q beta

validScalarCommitParams :: Commit.CommitParams -> Bool
validScalarCommitParams p = Commit.validCommitParams p && Commit.cp_n1 p == 1

validConfidentialCommitKey :: Commit.CommitParams -> [[Int]] -> Bool
validConfidentialCommitKey = Commit.separating_commit_key

randCommitKey :: Commit.CommitParams -> [[Int]] -> [[Int]]
randCommitKey p = map (drop (Commit.cp_n1 p))

randCommit :: Commit.CommitParams -> [[Int]] -> [Int] -> [Int]
randCommit p ck r =
  Zq.vec_mod (Listvec.mat_vec_mult (randCommitKey p ck) r) (Commit.cp_q p)

amountOfOpening :: Commit.CommitOpening -> Int
amountOfOpening op =
  case Commit.open_msg op of
    x : _ -> x
    [] -> 0

openingAdd :: Commit.CommitOpening -> Commit.CommitOpening -> Commit.CommitOpening
openingAdd op1 op2 =
  Commit.makeOpening
    (Listvec.vec_add (Commit.open_msg op1) (Commit.open_msg op2))
    (Listvec.vec_add (Commit.open_rand op1) (Commit.open_rand op2))

openingSub :: Commit.CommitOpening -> Commit.CommitOpening -> Commit.CommitOpening
openingSub op1 op2 =
  Commit.makeOpening
    (Listvec.vec_sub (Commit.open_msg op1) (Commit.open_msg op2))
    (Listvec.vec_sub (Commit.open_rand op1) (Commit.open_rand op2))

aggregateOpening ::
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  Commit.CommitOpening
aggregateOpening opIn1 opIn2 opOut1 opOut2 =
  openingSub (openingAdd opIn1 opIn2) (openingAdd opOut1 opOut2)

aggregateRandomness ::
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  Commit.CommitOpening ->
  [Int]
aggregateRandomness opIn1 opIn2 opOut1 opOut2 =
  Commit.open_rand (aggregateOpening opIn1 opIn2 opOut1 opOut2)

balanceCommitment :: [Int] -> [Int] -> [Int] -> [Int] -> Int -> [Int]
balanceCommitment cIn1 cIn2 cOut1 cOut2 q =
  Zq.vec_mod
    (Listvec.vec_sub (Listvec.vec_add cIn1 cIn2) (Listvec.vec_add cOut1 cOut2))
    q

validBalanceWitness :: Commit.CommitParams -> [Int] -> Bool
validBalanceWitness p r =
  Listvec.valid_vec (Commit.cp_n2 p) r &&
  Norms.all_bounded r (4 * Commit.cp_beta p)

validBalanceMask :: Commit.CommitParams -> Int -> [Int] -> Bool
validBalanceMask p gamma y =
  Listvec.valid_vec (Commit.cp_n2 p) y &&
  Norms.all_bounded y gamma

balanceResponseBound :: Commit.CommitParams -> Int -> Int -> Int
balanceResponseBound p gamma challenge =
  gamma + abs challenge * (4 * Commit.cp_beta p)

validBalanceChallenge :: Commit.CommitParams -> Int -> Bool
validBalanceChallenge p challenge =
  validScalarCommitParams p &&
  (challenge == 0 || challenge == 1)

validBalanceResponse :: Commit.CommitParams -> Int -> Int -> [Int] -> Bool
validBalanceResponse p gamma challenge z =
  Listvec.valid_vec (Commit.cp_n2 p) z &&
  Norms.all_bounded z (balanceResponseBound p gamma challenge)

balanceRelation :: Commit.CommitParams -> [[Int]] -> [Int] -> [Int] -> Bool
balanceRelation p ck c r =
  validScalarCommitParams p &&
  validConfidentialCommitKey p ck &&
  validBalanceWitness p r &&
  randCommit p ck r == c

balanceSigmaCommit :: Commit.CommitParams -> [[Int]] -> [Int] -> [Int]
balanceSigmaCommit = randCommit

balanceSigmaRespond :: [Int] -> [Int] -> Int -> [Int]
balanceSigmaRespond r y challenge =
  Listvec.vec_add y (Listvec.scalar_mult challenge r)

canonicalBalanceChallenge :: Commit.CommitParams -> [[Int]] -> [Int] -> [Int] -> Int
canonicalBalanceChallenge p ck c a =
  mod (sum (concat ck) + sum c + sum a) 2

balanceSigmaVerify ::
  Commit.CommitParams ->
  Int ->
  [[Int]] ->
  [Int] ->
  [Int] ->
  Int ->
  [Int] ->
  Bool
balanceSigmaVerify p gamma ck c a challenge z =
  validScalarCommitParams p &&
  validConfidentialCommitKey p ck &&
  Listvec.valid_vec (Commit.cp_m p) a &&
  validBalanceChallenge p challenge &&
  validBalanceResponse p gamma challenge z &&
  randCommit p ck z ==
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) (Commit.cp_q p)

balanceFsChallenges :: Commit.CommitParams -> [[Int]] -> [Int] -> [[Int]] -> [Int]
balanceFsChallenges _ ck c as =
  RepeatedFS.boolFsChallenges (sum (concat ck) + sum c + sum (concat as))

balanceSigmaResponds :: [Int] -> [[Int]] -> [Int] -> [[Int]]
balanceSigmaResponds r ys es =
  RepeatedFS.sigmaResponseRounds balanceSigmaRespond r ys es

balanceFsProve ::
  Commit.CommitParams ->
  Int ->
  [[Int]] ->
  [Int] ->
  [Int] ->
  [[Int]] ->
  Maybe BalanceProof
balanceFsProve p gamma ck c r ys =
  let as = map (balanceSigmaCommit p ck) ys
      es = balanceFsChallenges p ck c as
      zs = balanceSigmaResponds r ys es
      validMasks = length ys == balanceFsRounds &&
        and [validBalanceMask p gamma (ys !! i) | i <- [0 .. balanceFsRounds - 1]]
      validResponses = length zs == balanceFsRounds &&
        and [validBalanceResponse p gamma (es !! i) (zs !! i) | i <- [0 .. balanceFsRounds - 1]]
   in if balanceRelation p ck c r &&
         validMasks &&
         validResponses
        then Just (BalanceProof as zs)
        else Nothing

balanceFsVerify ::
  Commit.CommitParams ->
  Int ->
  [[Int]] ->
  [Int] ->
  BalanceProof ->
  Bool
balanceFsVerify p gamma ck c proof =
  let as = balance_as proof
      es = balanceFsChallenges p ck c as
      zs = balance_zs proof
   in length as == balanceFsRounds &&
      length zs == balanceFsRounds &&
      and
        [ balanceSigmaVerify p gamma ck c (as !! i) (es !! i) (zs !! i)
        | i <- [0 .. balanceFsRounds - 1]
        ]
